### PR TITLE
Refactor bemhtml

### PR DESCRIPTION
**Pull request by tadatuta**
// cc @tormozz48 @tavriaforever @gela-d

### CODE REVIEW

File: src/blocks/common.blocks/post/__title/post__title.bemhtml [hunk 1]
```diff
@@ -1,14 +1,24 @@
-block post, elem title {
+block('post').elem('title')(
+    def().match(function () {
+        return !this.ctx.content || this.ctx.content.length == 0;
```
Comment (gela-d) [inline src/blocks/common.blocks/post/__title/post__title.bemhtml]: Let's write strict comparison `===`

File: src/blocks/common.blocks/posts/posts.bemhtml [hunk 2]
```diff
@@ -1,14 +1,19 @@
...
-            elem: 'title',
-            content: this.ctx.title
-        },
-        applyNext()
-    ]
-}
+block('posts')(
+    def().match(function () {
+        var content = this.ctx.content;
+        return content && content.length == 0;
```
Comment (gela-d) [inline src/blocks/common.blocks/posts/posts.bemhtml]: `===`?

File: src/blocks/common.blocks/posts/posts.bemhtml [hunk 3]
```diff
@@ -1,14 +1,19 @@
...
-        {
-            elem: 'title',
-            content: this.ctx.title
-        },
-        applyNext()
-    ]
-}
+block('posts')(
+    def().match(function () {
+        var content = this.ctx.content;
```
Comment (gela-d) [inline src/blocks/common.blocks/posts/posts.bemhtml:3]: Can we add `var title = this.ctx.title;` ?

**Comment by gela-d** [comment]
Volodya, thank u for your help!
Just in case, we need all of these code in `leveldb`-tree

File: src/blocks/common.blocks/author/__avatar/author__avatar.bemhtml [hunk 4]
```diff
@@ -5,9 +5,10 @@ block('author').elem('avatar')(
     })(
         tag()('img'),
         attrs()(function() {
+            var content = this.ctx.content;
             return {
```
Comment (sipayRT) [inline src/blocks/common.blocks/author/__avatar/author__avatar.bemhtml:11]: add line between `var` and `return`

File: src/blocks/common.blocks/error-billboard/error-billboard.bemhtml [hunk 5]
```diff
@@ -1,6 +1,6 @@
-block error-billboard {
-
-    content: {
+block('error-billboard')(
+    content()(function () {
+        var code = this.ctx.code;
```
Comment (sipayRT) [inline src/blocks/common.blocks/error-billboard/error-billboard.bemhtml:3]: line between `var` and `return`

File: src/blocks/common.blocks/error-billboard/error-billboard.bemhtml [hunk 6]
```diff
@@ -25,53 +25,53 @@ block error-billboard {
...
 
-        content: BEM.I18N('error-billboard', 'title')
+    elem('title')(
+        tag()('h1'),
 
-        this.ctx.code, content: BEM.I18N('error-billboard', 'title') + ' ' + this.ctx.code
-    }
+        content()(function () {
+            var code = this.ctx.code;
+            return BEM.I18N('error-billboard', 'title') + (code ? (' ' + code) : '');
```
Comment (sipayRT) [inline src/blocks/common.blocks/error-billboard/error-billboard.bemhtml:38]: the same

File: src/blocks/common.blocks/post/__date/post__date.bemhtml [hunk 7]
```diff
@@ -18,18 +17,14 @@ block('post').elem('date')(
...
-        }
+        return ctx.editDate ? [
+            BEM.I18N('post', 'editDate'),
+            ' ',
+            format(ctx.editDate)
+        ] : [
+            BEM.I18N('post', 'createDate'),
+            ' ',
+            format(ctx.createDate)
+        ];
```
Comment (sipayRT) [inline src/blocks/common.blocks/post/__date/post__date.bemhtml:25]: maybe:

```
var actionType = ctx.editDate ? 'editDate' : 'createDate';

return [
    BEM.I18N('post', actionType),
    ' ',
    format(ctx[actionType])
];
```

?

File: src/blocks/common.blocks/post/__title/post__title.bemhtml (hunk 1 above)
Comment (sipayRT) [inline src/blocks/common.blocks/post/__title/post__title.bemhtml]: your pre-commit hook shouldn't give to commit such things

File: src/blocks/common.blocks/post/__title/post__title.bemhtml [hunk 8]
```diff
@@ -1,14 +1,24 @@
...
+    })(false),
 
-    default, (!this.ctx.content || this.ctx.content.length == 0): {
-        return false;
-    }
+    match(function () {
+        var mods = this.ctx.elemMods;
+        return mods && mods.level;
```
Comment (sipayRT) [inline src/blocks/common.blocks/post/__title/post__title.bemhtml:9]: new line everywhere

File: src/blocks/common.blocks/posts/posts.bemhtml (hunk 3 above)
Comment (sipayRT) [inline src/blocks/common.blocks/posts/posts.bemhtml:3]: no, it's another scope

**Comment by tavriaforever** [comment]
@tadatuta Are you test it?

File: src/blocks/common.blocks/post/__title/post__title.bemhtml (hunk 1 above)
Comment (gela-d) [inline src/blocks/common.blocks/post/__title/post__title.bemhtml]: @sipayRT we haven't yet pre-commit hook in this repository ;)

**Comment by gela-d** [comment]
This tree on bem-info is ok.

**Comment by gela-d** [comment]
I can't check our site with this commit, because of impossibility of cherry-pick this commit to leveldb.

**Comment by tadatuta** [comment]
up

**Comment by tadatuta** [comment]
@gela-d https://github.com/bem/bem-site-engine/pull/85

**Comment by gela-d** [comment]
Let's squash and merge

**Pull request merged**